tag.go

```diff
@@ -82,75 +82,165 @@ func (tag Tag) ElemCount() int {
  */
 func ParseQualifiedTagName(qtn string) ([]string, error) {
 	var ret []string
+	i := 0
 
-	if qtn == "" {
-		return nil, fmt.Errorf("Empty tagname supplied")
+	alpha := "abcdefhijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+	num := "0123456789"
+	alphanum := alpha + num + "_"
+
+	/* SYMBOLIC_SEG := [a-zA-Z][a-zA-Z0-9_]* */
+	parseSymbolicSegment := func() error {
+		var sb strings.Builder
+		begin := i
+
+		/* [a-zA-Z] */
+		if i >= len(qtn) {
+			return fmt.Errorf("Expected alphabetic character")
+		}
+		if unicode.IsSpace(rune(qtn[i])) {
+			return fmt.Errorf("Expected alphabetic character, got whitespace")
+		}
+		if !bytes.ContainsAny([]byte{qtn[i]}, alpha) {
+			return fmt.Errorf("symbolic sequence begins with a non-alphabetic character '%c'", qtn[i])
+		}
+		sb.WriteByte(qtn[i])
+
+		/* [a-zA-Z0-9_]* */
+		for ; i < len(qtn); i++ {
+			if unicode.IsSpace(rune(qtn[i])) {
+				return fmt.Errorf("Expected alphanumeric character, got whitespace")
+			}
+			if !bytes.ContainsAny([]byte{qtn[i]}, alphanum) {
+				ret = append(ret, qtn[begin:i])
+				return nil
+			}
+			sb.WriteByte(qtn[i])
+		}
+		ret = append(ret, qtn[begin:i])
+		return nil
 	}
-	insideBrackets := false
-	for i, c := range qtn {
-		if c == '[' {
-			if insideBrackets {
-				return nil, fmt.Errorf("Nested brackets at index %d", i)
+
+	/* NUMERIC_SEG := [:space:]* [0-9]+ [:space:]* */
+	parseNumericSegment := func() error {
+		var sb strings.Builder
+		var begin int
+
+		/* [:space:]* */
+		if i >= len(qtn) {
+			return fmt.Errorf("expected numeric character")
+		}
+		for ; i < len(qtn); i++ {
+			if !unicode.IsSpace(rune(qtn[i])) {
+				break
+			}
+		}
+
+		begin = i
+
+		/* [0-9]+ */
+		if i >= len(qtn) {
+			return fmt.Errorf("expected numeric character")
+		}
+		if !bytes.ContainsAny([]byte{qtn[i]}, num) {
+			fmt.Errorf("Expected digit, got '%c'", qtn[i])
+		}
+		sb.WriteByte(qtn[i])
+
+		for ; i < len(qtn); i++ {
+			if unicode.IsSpace(rune(qtn[i])) {
+				break
+			}
+			if !bytes.ContainsAny([]byte{qtn[i]}, num) {
+				goto parseAndExit
 			}
-			insideBrackets = true
+			sb.WriteByte(qtn[i])
 		}
-		if c == ']' {
-			if !insideBrackets {
-				return nil, fmt.Errorf("Mis-matched brackets at index %d", i)
+
+		/* [:space:]* */
+		for ; i < len(qtn); i++ {
+			if !unicode.IsSpace(rune(qtn[i])) {
+				break
```

Comment:
It looks like this loop is only executed after the if statement on line 147, so it could effectively be inside that if statement. Then the `goto` could become a `break`.